initial changes for invoker cluster backpressure

tysonnorris opened the pull request:
<!--- Provide a concise summary of your changes in the Title -->

## Description
<!--- Provide a detailed description of your changes. -->
<!--- Include details of what problem you are solving and how your changes are tested. -->

## Related issue and scope
<!--- Please include a link to a related issue if there is one. -->
- [ ] I opened an issue to propose and discuss this change (#????)

## My changes affect the following components
<!--- Select below all system components are affected by your change. -->
<!--- Enter an `x` in all applicable boxes. -->
- [ ] API
- [ ] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [ ] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes
<!--- What types of changes does your code introduce? Use `x` in all the boxes that apply: -->
- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [ ] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:
<!--- Please review the points below which help you make sure you've covered all aspects of the change you're making. -->

- [ ] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [ ] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [ ] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

chetanmeh commented on core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/ContainerPool.scala:
```diff
@@ -69,6 +75,20 @@ class ContainerPool(childFactory: ActorRefFactory => ActorRef,
+  //if cluster managed resources, subscribe to events
+  if (poolConfig.clusterManagedResources) {
+    logging.info(this, "subscribing to NodeStats updates")
+    Events.subscribe(self, NodeStatsUpdateEvent)
```
Should it unsubscribe itself? Given its kind of singleton its fine but may be for testcases we should unsubscribe

tysonnorris replied:
Looks like EventStream will handle unsubscribe on actor termination per docs, but will change this if it is a problem for tests.

chetanmeh commented on common/scala/src/main/scala/org/apache/openwhisk/utils/Events.scala:
```diff
@@ -0,0 +1,59 @@
+ * Publishes the payload of the MsgEnvelope when the topic of the
+ * MsgEnvelope equals the OWEvent specified when subscribing.
+ */
+object Events extends EventBus with LookupClassification {
```
We can possibly use the [EventStream][1] of the `ActorSystem` itself. Should meet our requirement

[1]: https://doc.akka.io/docs/akka/2.5/event-bus.html#event-stream

tysonnorris replied:
Sounds good! will look into making this change

chetanmeh reviewed (commented):
This would really address the dynamic capacity addition aspect! Do we have the change in MesosContainerFactory somewhere which trigger the `NodeStatsUpdateEvent`? That would help to see complete picture once

It may be better to extract out a `ResourcePolicy` abstraction to keep the policy aspect separate from the scheduling purpose

tysonnorris commented:
I'm interested to know what you have in mind for `ResourcePolicy` abstraction - it may help once I get the mesos PR setup, will do that today.

chetanmeh commented:
Was thinking of trait which abstracts the `hasPoolSpaceFor` method (like [here][1]). However current logic closely match the reservation and I am getting bit confused between its role and free and busy pool.

However idea would be to move your current changes to such a policy and then use that in ContainerPool decision logic. One aspect which makes it tricky is that policy need to be refreshed via event. So we may need to use some volatile state which gets updated after post processing of cluster state changes and then consumer by the `hasPoolSpaceFor` logic

[1]: https://github.com/adobe-apiplatform/incubator-openwhisk/commit/725f4b12ca85ff5cc4dd42a45fc8b2b289eb5ed9

tysonnorris committed:
"invoker clustered resources"

tysonnorris committed:
add a throttle for number of in-flight "container starts" when using cluster managed resources

tysonnorris committed:
remove tight loop in busy retries

tysonnorris committed:
ContainerPool must evict idles when certain portion of hosts are no longer available

tysonnorris committed:
ContainerPool must evict idles when certain portion of hosts are no longer available

tysonnorris committed:
ContainerPool must evict idles when certain portion of hosts are no longer available

tysonnorris committed:
refactoring to isolate cluster/local communications to ContainerResourceManager impls; handle ClusterResourceError in case of reaching cluster capacity

tysonnorris committed:
add GB to size

tysonnorris committed:
cleanup

tysonnorris committed:
cleanup

tysonnorris committed:
cleanup; increase replicated data sharing to once per second

tysonnorris committed:
added resent list of activationIds to avoid resending duplicates

tysonnorris committed:
added metrics for resource error, rescheduled activation, number nodes, total memory, max memory

tysonnorris committed:
review feedback

tysonnorris committed:
use histogram for node count

tysonnorris committed:
don't emit metrics from empty nodestats

tysonnorris committed:
tolerate empty nodestats

tysonnorris committed:
shuffle dependencies

tysonnorris committed:
shuffle dependencies

tysonnorris committed:
cleanup

tysonnorris committed:
cleanup; adding metrics

tysonnorris committed:
typo in conf

tysonnorris committed:
added tests; removed some methods defined in ContainerResourceManager

tysonnorris committed:
cleanup

tysonnorris committed:
use gauge metrics; update unused refs when containers are purged after idle timeout

tysonnorris committed:
change container pool metrics to gauge type, add metric for runbuffer size

tysonnorris committed:
fixing idle metrics

tysonnorris committed:
add counter metric for system errors

chetanmeh committed:
Update version for Akka dependencies and Scala (#4316)

mrutkows committed:
Additional cleanup and simplifications for the ActionLoop docs (#4475)

* Additional cleanup and simplifications for the ActionLoop docs

* Additional cleanup and simplifications for the ActionLoop docs

jthomas committed:
Updating runtimes to include new Node.js v12 image. (#4472)

* Updating runtimes to include new Node.js v12 image.

- Updated runtimes manifest
- Added API documentation
- Minor updates to docs
- Added automated test case

* Fixing accidental default flag for v12

Commit:
Update LeanBalancer for invoke SPI change (#4478)

KeonHee committed:
Limit the minimum to the cpu-share value (#4477)

Commit:
Adding Lean Openwhisk test to Jenkins (#4480)

style95 committed:
Update docker client version to 18.06.3 (#4430)

* Update docker version in the controller and guides
* Update travis configuration to install docker=18.06.3
* Separate docker installation from the travis setup script
* Add docker setup script
* Add CoordinatedShutdown to cleanup runtime containers
* add a configuration for root runc dir
* Disable runc use in Jenkins environment
* Add comments which explain the correlation among the version docker and runc and the type of user
* Reenable Docker remote API again

Commit:
Fixing Jenkins Lean Openwhisk test (#4482)

neerajmangal committed:
Invoker agent failed to pull docker images if using docker registry name (#4488)

Add `pip install docker` to Dockerfile for ow-utils to fix problem pulling docker images

chetanmeh committed:
Switch to consistent indexing policy for CosmosDB (#4484)

* Switch to consistent indexing policy
* Remove reference to Range index
* Tweak indexing policy comparison to only check for included and excluded path
Do not check for Index type as now there is only one which is Range for our cases
* Use implicit logger
* Excluding root path should be using `/*` instead of `/`

chetanmeh committed:
Update to Alpakka S3 Connector v1.0 release (#4479)

Updates to Alpakka S3 Connector v1.0.1 release.

This commit also includes some fixes related to #4484 which were causing compatibility issues with existing setup. For existing setup indexes were still using `Hash` indexing and that caused failure while creating `IndexingPolicy` upon collection read. So added back support for `Hash` index but not using them to create new indexes now

mhenke1 committed:
add quotes to avoid build brakes (#4491)

chetanmeh committed:
Track collection usage metrics for CosmosDB (#4490)

Record collection usage stats for CosmosDB so as to enable tracking the growth of collection in  terms of storage size, document count and index size over the period of time. It also enables tracking any indexing progress if any change is done in Index configuration.

Note that Count stats are currently not exposed via Azure Portal

Further this commit also enables emitting verbose trace for query when in debug mode. This would simplify any query performance analysis.

Fixes #4489

rabbah committed:
Apache OpenWhisk logos as svg (#4493)

chetanmeh committed:
Use `linesIterator` instead of String.lines (#4495)

`lines` method is now defined as part of java.lang.String in JDK 11. So need to use `linesIterator` for right method to be picked

chetanmeh committed:
Track Kafka client side metrics via Kamon (#4481)

Adds a configurable MetricsReporter to route Kafka metrics to Kamon once enabled. Set of metrics names which need to be captured needs to be explicitly configured

chetanmeh committed:
Update rest-assured to v4.0.0 (#4500)

* Update to restassured v4.0 which is compatible with jdk11

* Need to initialize both truststore and keystore for ssl cert to be validated

chetanmeh committed:
Use Instant with milli second precision (#4497)

dragos committed:
Fix a few compilation accidents (#4505)

* Add explicit return type on a few implicit values.

Implicit values without an explicit return types are not guaranteed to work. They may or may not compile, depending on compilation order and they are going to be disallowed in future versions of Scala. This confuses IntelliJ as well.

* Make Exec public since it's being used from tests.

The call to Exec.isBinaryCode in ActionContainer.scala is from a different package and normally not visible at the call site. Due to scala/bug#11554 this may compile sometimes.

rabbah committed:
Remove default port on api host. (#4504)

kiran002 committed:
Consolidated action annotations to a new singleton (#4499)

tysonnorris committed:
add idleGrace duration to prevent aggressive collection of recently idle containers; only remove idle containers that match the actual host/port of selected container

mhenke1 committed:
Settle throttle before running the API GW Tests (#4496)

* simplify throttle code

* revert to original throttle algorithm

* Make the waiting time to calm down the thottle explicit

rabbah committed:
Replace JsBoolean() with JsTrue/False. (#4508)

mrutkows committed:
Update all files using Mini license header to use full ASF header (#4512)

* Update Ansible YAML playbooks to use full ASF header

* Update remaining files to use full ASF header

tysonnorris committed:
respect idle grace for local remove() as well as remote triggered remove()

tysonnorris committed:
keep track of Container.lastUsed value properly to prevent early cleanup of containers

mrutkows committed:
Update Dockerfiles to full ASF license header (#4517)

selfxp committed:
Change API call to Splunk to allow retrieval of large messages (#4519)

Fixes #4518

sven-lange-last committed:
Log more diagnostic info when invoker / controller cluster status changes (#4520)

Whenever invoker or controller cluster status changes, the ShardingContainerPoolBalancer updates its state. If a new invoker registers, it is added to the load balancer status with its user memory capacity and a book-keeping semaphor is created to be used for scheduling load. If the controller cluster size changes, all invoker book-keeping semaphors are updated to reflect the user memory shard size managed by each controller.

This change logs more diagnostic information when the ShardingContainerPoolBalancer updates its state due to invoker or controller cluster size changes. The additional information helps with understanding how much user memory is available on invokers and is available on the controller's shard.

chetanmeh committed:
Openwhisk in a standalone runnable jar (#4516)

A standalone controller, using the lean load balancer and an in memory store.

* Use WhiskerControl port 3233 as default!
* Allow disabling read of whisk.properties, required for test run to avoid launched server influenced by whisk.properties generated for the tests. Also ensure that a system property is only set if no existing value found. This ensures that in special case user can still supersede the value being set via explicitly providing a system property override.
* Ensure that MemoryStore is only created when needed
* Fetch logs via `docker logs` command on Mac
* Add a OpenWhisk colored banner
* Package the default runtimes.json
* Include log timestamps and filter till sentinel
* Use host.docker.internal as api host for Mac and Windows
* Enable colored logging
* Add support for --verbose flag to enable debug logging easily
* Color transactionId and source also. Also support disabling color logging if needed
* Expose extension point for format log message
* Api host name is defined via system property
* Add a set of pre flight checks to confirm if OpenWhisk can run properly on this host
* Also check if docker is running
* New StandaloneDockerContainerFactory which adapts as per OS
* Only do pull for images having `openwhisk` prefix. For local image generally named `whisk/` no pull would be done
* Add ./gradlew :core:standalone:bootRun command to run jar directly. Use Spring boot default bootRun target to run the jar. Also copy the jar to bin directory.
* Make test run against standalone server
* Support disabling pull of standard images all together. Required for test runs
* Disable pause/resume support for non linux setups
* Print the wsk and docker cli version
* Update README with details on how to connect to db
* Increase time allowed for server to start to 30 secs and log logs upon timeout failure
* Enable verifySystemShutdown in Mesos tests

tysonnorris committed:
review feedback

tysonnorris committed:
akka version updates

tysonnorris committed:
fixing tests
